File: articles/calculus-for-systems-modeling/articles/sequences-series-and-the-logic-of-convergence/fortran/sequence_series_convergence_audit.f90
program sequence_series_convergence_audit
  implicit none
  real(8) :: geo, geo_ref, harm
  integer :: i

  geo = 0.0d0
  do i=0,24
    geo = geo + 10.0d0 * (0.6d0 ** i)
  end do
  geo_ref = 10.0d0 / (1.0d0 - 0.6d0)

  harm = 0.0d0
  do i=1,10000
    harm = harm + 1.0d0 / real(i,8)
  end do

  print '(A)', 'series_name n_terms last_term partial_sum reference_value estimated_error classification'
  print '(A,1X,I6,1X,F12.6,1X,F12.6,1X,F12.6,1X,F12.6,1X,A)', 'geometric_r_0.6',25,10.0d0*(0.6d0**24),geo,geo_ref,geo_ref-geo,'convergent'
  print '(A,1X,I6,1X,F12.6,1X,F12.6,1X,A)', 'harmonic',10000,1.0d0/10000.0d0,harm,'divergent'
end program

File: articles/calculus-for-systems-modeling/articles/sequences-series-and-the-logic-of-convergence/c/sequence_series_convergence_audit.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

double geometric_sum(double a, double r, int n){
  double total = 0.0;
  for(int i=0;i<n;i++){ total += a * pow(r, i); }
  return total;
}

double harmonic_sum(int n){
  double total = 0.0;
  for(int i=1;i<=n;i++){ total += 1.0 / i; }
  return total;
}

int main(void){
  double geo = geometric_sum(10.0, 0.6, 25);
  double geo_ref = 10.0 / (1.0 - 0.6);
  double harm = harmonic_sum(10000);
  printf("series_name,n_terms,last_term,partial_sum,reference_value,estimated_error,convergence_classification\n");
  printf("geometric_r_0.6,25,%.12f,%.12f,%.12f,%.12f,convergent geometric series\n",10.0*pow(0.6,24),geo,geo_ref,geo_ref-geo);
  printf("harmonic,10000,%.12f,%.12f,,,divergent despite terms approaching zero\n",1.0/10000.0,harm);
  return EXIT_SUCCESS;
}
